### ISSUE TITLE
Match Syntax and Precedence 

### ISSUE THREAD

**Issue by mariari** [heliaxdev/minijuvix#215]
Currently in MiniJuvix the following parses fine.

```rust
foo : ℕ -> ℕ;
foo x ≔ match x {
    (suc x) -> x;
    zero    -> suc zero;
  };
```

However, if I remove the `()`'s around the `suc x`:

```rust
foo : ℕ -> ℕ;
foo x ≔ match x {
    suc x -> x;
    zero  -> suc zero;
  };
```

then I get the following error message

```haskell
ile foo-new.mjuvix 
/home/katya/Documents/Work/repo/minijuvix/minijuvix-stdlib/foo-new.mjuvix:18:9: error:
/home/katya/Documents/Work/repo/minijuvix/minijuvix-stdlib/foo-new.mjuvix:18:9:
   |
18 |     suc x -> x;
   |         ^^
unexpected "x "
expecting "->" or '↦'

```

It seems that `()`'s are needed to match the precedence of definitions, however since we are using `->` notation, the `->` should go off last and parse the lhs as if it had ()'s over the outer form, much like function application:

In the language I write `suc x` not `(suc x)`.

Further if `suc` happended to be infix, likewise it should accept:
```rust
foo : ℕ -> ℕ;
foo x ≔ match x {
    blah :> y -> x;
    zero       -> suc zero;
  };
```

where it would be parsed as `(blah :> y) -> x`

**Comment by janmasrovira** [comment]
Indeed, `suc x` should not need a parenthesis on the pattern of a match.

After internal discussions we have decided to remove the `match` construct from the language. Instead, the standard library (or the user) will have a definition similar to this one:
```
 match : {A : Type} -> {B : Type} → A → (A → B) → B;
 match x f ≔ f x;
```
Then one will be able to write:
```
 if : {A : Type} → Bool → A → A → A;
 if c t f ≔ match c λ {
      true ↦ t;
      false ↦ f;
    };
```
which is very close to the currenty syntax.

Unfortunately, lambdas are not yet supported so it will take some time for this to be supported.

**Comment by mariari** [comment]
> Indeed, `suc x` should not need a parenthesis on the pattern of a match.
> 
> After internal discussions we have decided to remove the `match` construct from the language. Instead, the standard library (or the user) will have a definition similar to this one:
> 
> ```
>  match : {A : Type} -> {B : Type} → A → (A → B) → B;
>  match x f ≔ f x;
> ```
> 
> Then one will be able to write:
> 
> ```
>  if : {A : Type} → Bool → A → A → A;
>  if c t f ≔ match c λ {
>       true ↦ t;
>       false ↦ f;
>     };
> ```
> 
> which is very close to the currenty syntax.
> 
> Unfortunately, lambdas are not yet supported so it will take some time for this to be possible.

I think this works! At least I think the lazy behavior of `match` is preserved in the lambda call!

However now I have an unrelated question, will something similar happen to let bindings? At least the match here reminds me a lot of let expansion if we don't take let as a primitive and lambda encode it.
